lib/socket_proxy.ex

```diff
@@ -1,2 +1,42 @@
 defmodule SocketProxy do
+  @moduledoc """
+  Supervised process that spends most of its time in a TCP accept 
+  loop. If the accept loop fails, its supervisor should restart it all.
+  """
+
+  use GenServer
+  require Logger
+
+  def start_link({port, destinations}) do
+    GenServer.start_link(__MODULE__, {port, destinations})
+  end
+
+  def init({port, destinations}) do
+    Logger.info("Initializing Listener")
+    {:ok, lsock} = :gen_tcp.listen(port, [:binary, active: false, reuseaddr: true])
+    Logger.info("Listening on port #{port}")
+    send self(), {:start, lsock, destinations}
+    {:ok, lsock}
+  end
+
+  def handle_info({:start, lsock, destinations}, lsock) do
+    loop_accept(lsock, destinations)
+  end
+  def handle_info(unknown, state) do
+    Logger.info("Listener received unknown message: #{unknown}")
+    {:noreply, state}
+  end
+
+  defp loop_accept(lsock, destinations) do
+    case :gen_tcp.accept(lsock) do
+      {:ok, sock} ->
+        Logger.info("Accepted socket: #{Util.format_socket(sock)}")
+        spawn fn -> SocketProxy.Receiver.proxy(sock, destinations) end
+        loop_accept(lsock, destinations)
```

Comment:
You'll never leave the initial `handle_info` call this way. I think the pattern I've seen is accept the connection, then create a new child process to call `accept` again.

Reply:
That's true, but I couldn't think why it would be an issue. I've read that integrating the blocking TCP model into an OTP structure is sort of tricky. Can you give an example why being stuck in this loop is a problem? It would miss messages being sent, but the only messages I could think of would be from the supervisor to shut down, maybe? But then the supervisor would brutal kill it after a little bit anyway.

The `accept` could certainly be done in a child process (that's the way I had it initially, actually), but then it gets more complex handling if the listening socket has an issue. How would you handle the listen socket closing, in order to start listening on it again? The child could send a message up, but I feel like it starts to get into "no obvious error" territory as opposed to "obviously no errors". 

[This erlang email list question](http://erlang.org/pipermail/erlang-questions/2008-February/032912.html) on the topic says that in a larger system, blocking is an issue but in a standalone app it wouldn't be. However, the code sample in it does have a nice way of handling it, by using `listen/2` which has a timeout, and then re-listening. Essentially the listen loop becomes something like this instead:

```ex
def init(...) do
  # ...
  GenServer.cast(self(), :accept)
  {:ok, state}
end

def handle_cast(:accept, _from, state) do
  case :gen_tcp.accept(state.lsock, 3_000) do
    {:ok, sock} -> # ...
    {:error, :timeout} ->
      GenServer.cast(self(), :accept)
      {:noreply, state)
    {:error, error} -> # ...
  end
end
```

What do you think of this approach? It's a little more complex than the simple infinite listen loop that I have now, but it allows the process to break out of its loop and receive OTP messages and the like, while still keeping this process solidly in control of the listen socket, which I like.

Reply:
I'm not sure you can `cast` to yourself, but other than that I think it's definitely nicer. I think it's going to be less confusing if there's a timeout or something, and the next person to look at this code is wondering why no other messages are handled.

Reply:
I'd also recommend having the listening children be supervised: maybe a `simple_one_for_one` supervisor you can pass the accepted socket to?

Reply:
Okay, made the change to self-cast. (It worked just fine, I think the only reason we normally do the self-send and `handle_info` is to use `Process.send_after`.)

I'm actually ambivalent about supervising the socket receiver. I was trying to think of a scenario in which I'd want the supervisor to actually restart the child and I couldn't. The most likely way for the child to fail is if the socket closes or has some issue. But if that's the case, then I don't want it to try to restart and keep listening on the dodgy socket; I want it to die, as well as all its `spawn_link`'ed children. The OCC server will reconnect to opstech3, and the accept loop will pick it up and kick off a new child, which will create new connections to AWS.

So it's kinda/sorta supervised already if you squint - if the process dies, the upstream infrastructure will reconnect and spawn new processes.

Can you think of a reason why we'd want to respawn these processes with the socket in its state? I suppose, if not, at the very least I should at better documentation about the reasoning here.

Reply:
It's less about restarting if they crash, but rather about being able to cleanly shut them down if something else crashes. If you're using `spawn/1` to create new processes, they're free-floating rather than being part of a tree. You can give the supervised processes a restart strategy of `:temporary` to not have the supervisor restart them if they crash.

Reply:
Hmm, that makes sense. In this case, there's not much danger of them becoming free floating processes, I think, since they're stuck in a receive loop which exits (and takes down the `spawn_link`'ed processes they're sending messages to) if the socket is broken.

That said, I suppose for a robust OTP application you may want to be able to kill the processes even if the socket is healthy.

I updated the code to use a `:simple_one_for_one` supervisor, but now I have another dilemma about the process receiving its shutdown messages during the TCP `recv` loop.

The most efficient way to listen to the socket is to call `gen_tcp.recv/2`, which blocks until you get data, then handles the data, and loop and call `gen_tcp.recv/2` again. But doing that means the process won't be listening for any incoming messages. No problem - we can just do `use GenServer, shutdown: :brutal_kill`, and the supervisor will kill it dead if it needs to.

It's not super elegant, but it works and is low overhead.

Alternatively, we could go with a cast-self-and-handle approach, like we're doing with the accept loop. Then the process will mix in any normal messages it gets, but at the cost of basically having to send itself a message with every TCP packet, which seems like a lot of overhead to me.  We only get ~10 OCS messages/second so it _can_ keep up, I suppose, but I don't especially like generating unnecessary churn.

For now, I've gone with the quick-and-dirty `shutdown: :brutal_kill` approach, but could be convinced to introduce the message loop if you think it's better.

Reply:
You can also use [active mode](http://www.erlang.org/doc/man/inet.html#setopts-2) instead of explicitly calling `recv`:

> If the value is true, which is the default, everything received from the socket is sent as messages to the receiving process.

Reply:
Sigh, yeah. I guess I'll do that. I've sort of come full circle. On `master` right now is a fully supervised tree of processes that listen in active mode! It's just that it didn't work right, since I had assumed there'd only be one incoming connection from the OCS server at a time, when really there are 3 or 4. I also saw that when it was running it was taking around ~8% CPU, which isn't a lot, but seemed high for such a little app.

So I re-wrote it to be a small little thing without the OTP structure and a simple infinite accept loop which spins off processes in infinite recv loops.

But I guess it's better to do it the OTP way, and by fits and starts here we are again, but this time it will accept multiple incoming connections.

Just gotta update some tests now...